### PR TITLE
feat(clp-s): Add support for kv-pair-IR ingestion.

### PR DESCRIPTION
**Pull request by AVMatthews**
# Description

This pull request is the continuation or PR #618, adding the functionality promised. This adds the corresponding ingestion functionality to compress IRv2 into archive. 

- Add additional public function to JsonParser to expose parsing from IR; `parse_from_ir()`
- Add an class implementing the `clp::ffi::ir_stream::IrUnitHandlerInterface` for use deserializing IRv2 files; `IrUnitHandler`
- Add function to iterate over pairs in the `KeyValuePairLogEvent` converting the log event into the correct format for archive; `parse_kv_log_event()`
- Add two helper functions to help convert IR schema nodes into archive schema nodes:
    - Find the archive schema node type based on the IR schema node type and the value of the node in the current log event; `get_archive_node_type`
    - Find or add the correct archive schema node id for a given IR schema node id and correct archive schema node type for the value of the node in the current log event; `get_archive_node_id`

# Validation performed

For our posted JSON Datasets:
- Generated the corresponding IRv2 file
- Compress the IRv2 file to archive 
    - `./clp-s c <archive dir> <ir file or directory w/ ir file> --file-type kv-ir`
- Extract archive back to JSON
    - `./clp-s x <archive dir> <output JSON dir>`
- Sort keys and compare to JSON generated by clp-s from compressing and extracting directly from original JSON (spark, elasticsearch, and postgresql)
    - `jq -S -c '.' <output JSON dir>/original > out.json`
        - You can sort the json order if not using built in log ordering
        - `jq -S -c '.' <output JSON dir>/original | sort > out_sorted.json`
    - `diff -u out.json clp-s_out.json`
        - To generate clp-s_out.json:
            - `./clp-s c <archive dir> <original JSON>`
            - `./clp-s x <archive dir> <output JSON dir>`
            - `jq -S -c '.' <output JSON dir>/original > clp-s_out.json`
- Spot checked match of mongodb and cockroach database

- Notes:
    - I noticed some differences in numbers of significant digits for floats between both clp-s and the IRv2 to archive, and the original JSON.
    - ex.) original: 1.0; clp_s: 1.00000; IRv2: 1.00000
    - ex.) original: 0.9758767; clp_s: 0.9758767; IRv2: 0.975877

## Summary by CodeRabbit

- **New Features**
	- Enhanced `JsonParser` to support parsing of Key Value IR Streams.
	- Introduced new functionalities for handling log events and schema trees.
	- Improved error handling in the `compress` function for parsing Key Value IR files.
	- Added validation checks in command-line argument parsing for better user feedback.

- **Bug Fixes**
	- Enhanced error logging for parsing failures in the `compress` function.

### CODE REVIEW

**Review by gibber9809 — CHANGES_REQUESTED** [review]
Added an initial set of comments. Still need to go through `parse_from_ir` onwards, but should be enough comments to iterate on for now.

The other high level question I have is whether the `--timestamp-key` argument is supported for this ingestion path? If not we should throw in the same way we do if the `--structurize-arrays` argument is passed for IR ingestion.

File: components/core/src/clp_s/JsonParser.cpp
Comment (AVMatthews) [reply components/core/src/clp_s/JsonParser.cpp:49]: The deserializer expects a handler like the one laid out in components/core/src/clp/ffi/ir_stream/IrUnitHandlerInterface.hpp. In the test case for the deserializer the unused functions returned success like this, so I just followed suit. I'm not sure what the goal of the functions were. I think it exposed flexibility where if we wanted to handle each schema node one at a time, do something specific with each event, etc we could. Really we'd have to consult Zhihao to see if he had a specific thing in mind. All functions have pathways to get called by the deserializer, so we need to offer them. Is there something I'm supposed to used in the archive or parser with a UTC change?

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [reply components/core/src/clp_s/JsonParser.cpp:689]: Will leave it up to you whether to do this change right now since it's maybe a bit of a more complicated rewrite, but we should at least keep it in mind for a follow-up PR.

File: components/core/src/clp_s/JsonParser.cpp
Comment (AVMatthews) [reply components/core/src/clp_s/JsonParser.cpp:669]: 1.  The hash function used by the unordered map can't handle hashing a pair. We could write a custom hash function, but I could use some direction with what is the best practice for a custom hash function to avoid conflicts. The vector should be at the worst case 2 long, since in the worst case a ir schema node could translate to two different archive types. Originally I had a map<tuple<ir node id, NodeType>, archive node id>, but I was advice to change to this for better performance. 

2.  Since it's only used for the IR->archive pathway I'm inclined to leave it as is.  As for the name I kind of like having the structure type/ purpose recognized in the name, because when it's used I think it makes the action read a bit more like a sentence. Maybe ir_node_to_archive_node_id_translations or ir_node_to_archive_node_id_mapping?

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [reply components/core/src/clp_s/JsonParser.cpp]: For this and other utf8 validation cases we decided offline to leave it as is for now unless the escaping changes get merged before this PR is complete.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:728]:
```suggestion
    while (false == ir_id_stack.empty()) {
```
I think our coding standard asks for explicit evaluation against false.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:658]:
```suggestion
```
I think the thrown exception already gives enough information in this case.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:655]:
```suggestion
            if (node_has_value && node_value->is_null()) {
                return NodeType::NullValue;
            }
```

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:633]: If we're doing this check does node_value still need to be wrapped in std::optional?

Seems like we could rewrite this bit as `auto &node_value = kv_pair.second()`

And replace instances of node_has_value with `node_value.has_value()`.

Either that or declare node_value as `clp::ffi::Value` directly without the optional.

**Review by gibber9809 — CHANGES_REQUESTED** [review]
Mostly smaller changes this time.

Still want to confirm about the `--timestamp-key` behaviour I asked about in the last round of review.

Also wondering how we're tracking original size of the IR stream for metadata purposes. I didn't see anything in our code, and I'm guessing that the ir stream itself doesn't provide anything, but we should probably at least use an estimate like number of decompressed bytes consumed by the Decompressor or something.

File: components/core/src/clp_s/JsonParser.cpp
Comment (AVMatthews) [reply components/core/src/clp_s/JsonParser.cpp:844]: I guess I could also just continue onto the next file rather than returning. What do you think? I'll leave it as is for now, but would be a quick change to remove or to change to continue instead of return.

Comment (gibber9809) [reply components/core/src/clp_s/JsonParser.cpp:840]: Let's remove it then.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:847]:
```suggestion
        size_t curr_pos{};
        size_t last_pos{};
```

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:896]:
```suggestion
m_archive_writer->close();
```
No need to update metadata on the failure path. Also missed it earlier but we should close the archive on the error path to be consistent with JSON ingestion behaviour.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:872]:
```suggestion
                m_archive_writer->close();
                decompressor.close();
                return false;
```
Sorry, hadn't noticed this previously, but we should fail ingestion on errors from the IR deserializer.

File: components/core/src/clp_s/JsonParser.cpp
Comment (gibber9809) [inline components/core/src/clp_s/JsonParser.cpp:914]:
```suggestion
                m_archive_writer->close();
                decompressor.close();
                return false;
```
Sorry, missed this previously again, but we should error out on any IR unit we don't know how to handle. Want to avoid cases where we silently don't handle some feature and ingest an IR stream incorrectly.

For example, after reading through the IRv2 serializer/deserializer code a bit, it seems like we don't handle the UTC offset change feature. If we allow ingestion of streams that have changing UTC offsets then we'll create archives with incorrect timestamps.

**Review by gibber9809 — CHANGES_REQUESTED** [review]
A few more minor changes. Just confirming whether it is required to support UtcOffsetChange/expected behaviour for handling it. Should be good to go after that.

File: components/core/src/clp_s/JsonParser.cpp
Comment (AVMatthews) [reply components/core/src/clp_s/JsonParser.cpp:914]: This case covers both SchemaTreeNodeInsertion and UtcOffsetChange IR unit types. The SchemaTreeNodeInsertion is perfectly fine. we just don't need to do anything on it. I'll special case that one and fail in the else case.

**Review by gibber9809 — APPROVED** [review]
LGTM. For PR title how about `feat(clp-s): Add support for kv-pair-IR ingestion.`?

**Pull request merged**
